src/v1/parser.rs

```diff
@@ -0,0 +1,814 @@
+        })
+    }
+
+    /// Returns version of the signature file
+    pub fn get_version(&self) -> &str {
+        &self.version
+    }
+
+    /// Returns hashing algorithm
+    pub fn get_hash_type(&self) -> HashType {
+        self.hash_type
+    }
+
+    /// Returns block size
+    pub fn get_block_size(&self) -> u64 {
+        self.block_size
+    }
+}
+
+#[derive(Debug)]
+pub struct Footer(Vec<u8>);
+
+impl Footer {
+    fn parse(row: &[u8], hash_type: HashType)
+        -> Result<Footer, ParseRowError>
+    {
+        let digest_len = hash_type.get_size();
+        let hash_len = digest_len * 2;
+        if row.len() != hash_len {
+            return Err(ParseRowError(
+                format!("Footer length is {}, expected {}",
+                    row.len(), hash_len)));
+        }
+        let mut data = Vec::with_capacity(digest_len);
+        parse_hash(row, hash_len, &mut data)?;
+        Ok(Footer(data))
+    }
+}
+
+/// Entry hashes iterator
+#[derive(Debug, PartialEq)]
+pub struct Hashes {
+    data: Vec<u8>,
+    hash_type: HashType,
+}
+
+impl Hashes {
+    fn parse(row: &[u8], hash_type: HashType)
+        -> Result<Hashes, ParseRowError>
+    {
+        let digest_len = hash_type.get_size();
+        let hash_len = digest_len * 2;
+        let hashes_num = (row.len() + 1) / (hash_len + 1);
+        if row.len() != 0 && (row.len() + 1) % (hash_len + 1) != 0 {
+            return Err(ParseRowError(
+                format!("Hashes string length is {}, one hash should have exactly {} characters",
+                    row.len(), hash_len)));
+        }
+        let mut data = Vec::with_capacity(hashes_num * digest_len);
+        for hash in row.chunks(hash_len + 1) {
+            parse_hash(&hash[..hash_len], hash_len, &mut data)?;
+            if hash.len() == hash_len + 1 && hash[hash_len] != b' ' {
+                return Err(ParseRowError(
+                    format!("Invalid hashes delimiter: {:?}",
+                        String::from_utf8_lossy(&hash[hash_len..hash_len + 1]))));
+            }
+        }
+
+        Ok(Hashes {
+            data: data,
+            hash_type: hash_type,
+        })
+    }
+
+    fn check_size(&self, file_size: u64, block_size: u64)
+        -> Result<(), ParseRowError>
+    {
+        let expected_hashes_num = ((file_size + block_size - 1) / block_size) as usize;
+        if self.len() != expected_hashes_num {
+            return Err(ParseRowError(
+                format!("Expected {} hashes but found {}",
+                    expected_hashes_num, self.len())));
+        }
+        Ok(())
+    }
+
+    pub fn len(&self) -> usize {
+        self.data.len() / self.hash_type.get_size()
+    }
+
+    pub fn iter<'a>(&'a self) -> Chunks<'a, u8> {
+        self.data.chunks(self.hash_type.get_size())
+    }
+}
+
+/// Represents an entry from dir signature file
+#[derive(Debug)]
+pub enum Entry {
+    /// Direcory
+    Dir(PathBuf),
+    /// File
+    File(PathBuf, bool, u64, Hashes),
```

Comment:
Should be `File { filename: PathBuf, executable: bool, size: u64, ... }` otherwise it's hard to guess what these fields mean.